components/mhz19b/mhz19b.c

```diff
@@ -0,0 +1,357 @@
+    dev->buf = (uint8_t *) malloc(MHZ19B_SERIAL_BUF_LEN);
+    dev->last_value = -1;
+    dev->last_ts = esp_timer_get_time();
+    return ESP_OK;
+}
+
+esp_err_t mhz19b_free(mhz19b_dev_t *dev)
+{
+    free(dev->buf);
+    dev->buf = NULL;
+    return ESP_OK;
+}
+
+bool mhz19b_detect(mhz19b_dev_t *dev)
+{
+    uint16_t range;
+    // Check valid PPM range
+    if ((mhz19b_get_range(dev, &range) == ESP_OK) && (range > 0)) {
+        return true;
+    }
+
+    // Sensor not detected, or invalid range returned
+    // Try recover by calling setRange(MHZ19B_RANGE_5000);
+    return false;
+}
+
+bool mhz19b_is_warming_up(mhz19b_dev_t *dev, bool smart_warming_up)
+{
+    // Wait at least 3 minutes after power-on
+    if (esp_timer_get_time() < MHZ19B_WARMING_UP_TIME_US)
+    {
+        if (smart_warming_up)
+        {
+            ESP_LOGI(TAG, "Using smart warming up detection ");
+
+            int16_t co2, last_co2;
+            last_co2 = dev->last_value;
+            // Sensor returns valid data after CPU reset and keep sensor powered
+            mhz19b_read_CO2(dev, &co2);
+            if ((last_co2 != -1) && (last_co2 != co2)) {
+                // CO2 value changed since last read, no longer warming-up
+                return false;
+            }
+        }
+        // Warming-up
+        return true;
+    }
+
+    // Not warming-up
+    return false;
+}
+
+/*!
+ * \brief Check minimum interval between CO2 reads
+ * \details
+ *      Not described in the datasheet, but it is the same frequency as the built-in LED blink.
+ * \retval true
+ *      Ready to call readCO2().
+ * \retval false
+ *      Conversion not completed.
+ */
+bool mhz19b_is_ready(mhz19b_dev_t *dev)
+{
+    // Minimum CO2 read interval (Built-in LED flashes)
+    if ((esp_timer_get_time() - dev->last_ts) > MHZ19B_READ_INTERVAL_MS) {
+        return true;
+    }
+
+    return false;
+}
+
+/*!
+ * \brief Read CO2 from sensor
+ * \retval <0
+ *      MH-Z19B response error codes.
+ * \retval 0..399 ppm
+ *      Incorrect values. Minimum value starts at 400ppm outdoor fresh air.
+ * \retval 400..1000 ppm
+ *      Concentrations typical of occupied indoor spaces with good air exchange.
+ * \retval 1000..2000 ppm
+ *      Complaints of drowsiness and poor air quality. Ventilation is required.
+ * \retval 2000..5000 ppm
+ *      Headaches, sleepiness and stagnant, stale, stuffy air. Poor concentration, loss of
+ *      attention, increased heart rate and slight nausea may also be present.\n
+ *      Higher values are extremely dangerous and cannot be measured.
+ */
+esp_err_t mhz19b_read_CO2(mhz19b_dev_t *dev, int16_t *co2)
+{
+    esp_err_t result;
+    // Send command "Read CO2 concentration"
+    result = mhz19b_send_command(dev, MHZ19B_CMD_READ_CO2, 0, 0, 0, 0, 0);
+
+    // Check result
+    if (result == ESP_OK) {
+        // 16-bit CO2 value in response Bytes 2 and 3
+        *co2 = (dev->buf[2] << 8) | dev->buf[3];
+        dev->last_ts = esp_timer_get_time();
```

Comment:
Please explicitly include <esp_timer.h>. This works now because the esp_timer.h is included via esp_log.h, but this may change in future versions of ESP-IDF.